### PR TITLE
fix(types): allow usage of router option in InstantSearchConfig.routing

### LINKED ISSUE #390: No way to define routing.router function
https://github.com/algolia/angular-instantsearch/blob/de2a85ac695f5569cd83f8500c3b304572e2f40b/src/instantsearch/instantsearch.ts#L178

Our aim to to use SEO friendly URLs with a custom titles for different filters as mentioned here (https://community.algolia.com/instantsearch.js/v2/guides/routing.html#seo-friendly-urls)
But, unfortunately there is no way to define router settings as well as to use instantsearch.routers.history, because for routing we can only define stateMapping.

Is there is a way to define router settings for routing?

### PR DESCRIPTION
**Pull request by tkrugg**
Motivation:  
- allow usage of router option in InstantSearchConfig.routing  (fixes #390)

Drawbacks: None

Open questions: 
- are there any tests that can be added for this?
- shall we illustrate usage in the router example? I don't want overload it

### CODE REVIEW

**Comment by tkrugg** [comment]
@Haroenv 
> it’s better than before, but angular in this version still supports the Boolean version

InstantSearchJS support a boolean as well according to the examples in the [doc](https://community.algolia.com/instantsearch.js/v2/guides/routing.html#simple-url-synchronization).
It is true however that in the API ref [RouterOptions](https://community.algolia.com/instantsearch.js/v2/instantsearch.html#struct-RoutingOptions) can't be a boolean.
Either way removing the boolean support here would introduce a breaking change.

> The object type should be more detailed with the functions and their types

I think I know what you mean but I don't think we can do better because InstantSearchJS has no typescript types (so we can't infer the type of the object returned by `instantsearch.routers.history()`.

File: src/instantsearch/instantsearch.ts [hunk 1]
```diff
@@ -178,6 +178,13 @@ export type InstantSearchConfig = {
   routing?:
     | boolean
     | {
+        router?: {
+          onUpdate?: (cb: (object) => void) => void;
+          read?: () => object;
+          write?: (routeState: object) => void;
+          createURL?: (routeState: object) => string;
+          dispose?: () => void;
```
Comment (samouss) [inline src/instantsearch/instantsearch.ts]: My bad but the attributes are required, inside InstantSearch we don't assert them before calling them. `router` & `stateMapping` [are optionals](https://github.com/algolia/instantsearch.js/blob/a35cb0b7bc0cfbd83fa98648db31f0a6d002dd2f/src/lib/InstantSearch.js#L154-L159) but not their attributes.

File: src/instantsearch/instantsearch.ts [hunk 2]
```diff
@@ -179,11 +179,11 @@ export type InstantSearchConfig = {
...
-          onUpdate?: Function;
-          read?: Function;
-          write?: Function;
-          createURL?: Function;
-          dispose?: Function;
+          onUpdate?: (cb: (object) => void) => void;
+          read?: () => object;
+          write?: (routeState: object) => void;
+          createURL?: (routeState: object) => string;
+          dispose?: () => void;
```
Comment (tkrugg) [inline src/instantsearch/instantsearch.ts]: we said we'd make them all optional but I'm not so sure of it.
If I read the code correctly [the default router value is entirely overwritten by the config passed](url)
So for example if user doesn't define `onUpdate` this [line](https://github.com/algolia/instantsearch.js/blob/9632ade6434fb9257bdb23c66dde8aaa1c31c4eb/src/lib/RoutingManager.js#L52) will fail.

So making onUpdate optional is misleading

Comment (samouss) [reply src/instantsearch/instantsearch.ts]: Yep we comment at the same time see https://github.com/algolia/angular-instantsearch/pull/393#discussion_r239430417

File: src/instantsearch/instantsearch.ts (hunk 1 above)
Comment (tkrugg) [reply src/instantsearch/instantsearch.ts]: 👍 
I write a test as well, setting this PR as WIP

Comment (tkrugg) [reply src/instantsearch/instantsearch.ts]: I haven't find a way to test theses interfaces other than actually implementing an example with custom routing. https://github.com/algolia/angular-instantsearch/pull/395

**Pull request merged**